### DISCUSSION TITLE
Observations and Advice

### DISCUSSION THREAD

**Discussion by TCIII** [acredsfan/autonomous_mower#4] [General]
@acredsfan,

Please respond to my "Observations and Advice" in the [Issues](https://github.com/acredsfan/autonomous_mower/issues) Tab.

Regards,
TCIII

**Comment by TCIII** [comment]
@acredsfan,

I have added a number of additional posts to issue #1: "Observations and Advice".
I am presently trying to determine how to replace your `motor_controller.py` part with the DC [robohat.py](https://github.com/autorope/donkeycar/blob/main/donkeycar/parts/robohat.py) part and your `gps_interface.py` part with the DC [gps.py](https://github.com/autorope/donkeycar/blob/main/donkeycar/parts/gps.py) part so that I can use the new DC RC HAT with your Python code. 

Regards,
TCIII

**Comment by TCIII** [comment]
@acredsfan,

Have you had a chance to produce a new electrical schematic and take a look at some of my recent component suggestions?

Regards,
TCIII

**Reply by acredsfan** [reply]
Hi @TCIII,

I'm sorry, I have not been able to get the schematic together yet, but I will try to make time for that in the next few days.  I have looked at your component suggestions and have a BNO085 on hand as well, but am also interested in the RM3100.  Is there a need for both, or is the RM3100 your suggested replacement overall for the MPU9250?  I also received my Neo-9M module and hope to play around with that soon.   I appreciate the advice/heads up on RPI.GPIO, I wonder if that's been part of what my issues were with the motor drivers.

I also appreciate the links to the tutorials and information, it's really helpful!

Thanks,
acredsfan

**Comment by TCIII** [comment]
@acredsfan,

Thanks for the response, much appreciated.

I realize that Family comes first, but if we could map out a path forward, maybe I could work on some future changes/updates while waiting for you to become available.

For a start, do you think that you could adapt the Donkey Car [robohat.py part](https://github.com/autorope/donkeycar/blob/main/donkeycar/parts/robohat.py) in place of your motor_controller.py part? That way I could begin to move forward with your autonomous mower project. You could create a branch to store your modified motor controller version.

The BNO085, which is both an IMU and Magnetometer, can replace the MPU9250 IMU function, while the RM3100, which is only a Magnetometer, can replace the MPU9250 Magnetometer function. The BNO085, as an IMU, can be placed just about anywhere on the lawnmower chassis, while the RM3100 Professional Grade Magnetometer, like all Magnetometers, should be placed well away from soft/hard iron sources and, obviously, stationary/rotating magnetic fields whenever possible. Using a separate Magnetometer allows the Magnetometer to be placed independently of the IMU.

If you are using a RPi 4B 2GB, which I believe is no longer available, then I don't think that your motor control issues were due to a `RPI.GPIO` library compatibility issue. The `RPI.GPIO` and `pigpio` library issues appear to only affect the RPi 4B version 1.5 which were mainly RPi 4B 8GB modules.

Always glad to have been of help.

Regards,
TCIII

**Reply by acredsfan** [reply]
@TCIII ,

Absolutely and I appreciate your patience with me.  Looking back to the last notes I could recover from my old computer where I last was working on the robot, I was trying to debug the camera handling via the Singleton camera.  I was trying to use the picam as both an obstacle detection mechanism and to allow the user to track progress on the Web UI.  I had also repurposed the 2gb raspberry pi 4 to upgrade my Ender 5 plus to Klipper.  I'm currently using an 4GB pi 4b.  I was on the latest pi OS version as of late fall last year.  I was also attempting to use web sockets for lower latency on the steam and sensor data.  I have a bad habit of trying to change too many things close together. 😀

As far as a plan of attack, I'd like to get to a place where the wheel motors can be operated reliably and sensors working consistently first.  Your point about the GPIO problems on the pi 4b seems to mean there is some work to do there.

That being said, I'd defer to your expertise on direction, this is my first major robotics attempt so I'm still learning.  Feel free to make any changes you feel are appropriate.

I'm good with the sensors you've mentioned swapping, using RTK with the NEO-9M, and I'd like to be able to use the IBT-2 drivers as well as your robot HAT.  Ultimately, your DC HAT is the direction I think best fits the application from what you've shared, it's just ensuring options for both due to current availability.

Again, I really appreciate your expertise and engagement!

acredsfan

**Reply by TCIII** [reply]
@acredsfan,

Thanks for the insights, much appreciated.

I am a retired System Engineer so my main expertise is the integration of the software with the hardware.
I can somewhat read Python code, but am still learning its structure and how all the pieces of code go together.
If you create a path forward then you can identify the most important software changes and implement them one at a time until the whole system plays well together.

Since you have moved up to a RPi 4B 4GB, I recommend that you take a look at your RPi module and see if you can identify what version it is. RPi 4B  version 1.5 has issues with both the RPI.GPIO and pigpio libraries when running the Bookworm OS. You can search the IoT to see what those issues are.

As far as using the NEO-M9N as a RTK Rover, that unfortunately is a no-go as the NEO-M9N is not equipped to accept and integrate RTCM correction messages to improve its localization position.
This Sparkfun[ ZED-F9P RTK GPS module](https://www.amazon.com/gp/product/B07NBPNWNZ/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&psc=1) and the NavSpark Store PX1122R-EVB : [PX1122R L1/L2 RTK EVALUATION BOARD](https://navspark.mybigcommerce.com/px1122r-evb-px1122r-l1-l2-rtk-evaluation-board/), with an appropriate L1/L2 antenna, can be used for RTK GPS with either a Base Station or a NTRIP server.

Yes, the DC RC HAT can be used to control the autonomous lawnmower motors with or without a RC transmitter.

It  looks like [tlj1899](https://github.com/tlj1899) would like to be a collaborator. Check my Observations and Advice issue.

Regards,
TCIII

**Reply by acredsfan** [reply]
@TCIII

My bad on the NEO-M9N, I was confusing myself.  Overall, though, I still think this option along with an option of just using the M8N would be a good idea.  I'm ordering one of the modules you've linked here, is there a major downside to the cheaper PX112R-EVB that would make the price of the ZED-F9P worth the difference?

Thanks,
acredsfan

**Reply by TCIII** [reply]
@acredsfan,

All of the successful articles I have read on building an autonomous robotic lawnmower do not use either the NEO-M9N or the NEO-M8N due to their poor horizontal accuracy of 2.0 - 2.5 meters. On a good day a Rover RTK GPS module can get down to a horizontal accuracy of under two inches. I have been using this [Beitian BT-160 L1/L2 antenna](https://www.amazon.com/gp/product/B0BW8M9JR6/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&psc=1), which unfortunately is not currently available now, with my RTK GPS Rover modules. Last Christmas I paid $58 USD for my first BT-160, but after Christmas the price went up to $78 USD which is still a very good price. A decent alternative L1/L2 GPS antenna is this [Sparkfun antenna](https://www.amazon.com/gp/product/B09ZH8XWGK/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&psc=1). I think that you are going to find that you get what you pay for in the way of horizontal accuracy.

As far as the PX112R-EVB goes, the down side is the FedEx shipping cost from Taiwan which cost me $55 USD as I live in Florida. It might be a little less since you live in Ohio. I have both the Evaluation Board, which I am using in a Base Station and the NS-HP-GN2 : [PX1122R L1/L2 RTK BREAKOUT BOARD](https://navspark.mybigcommerce.com/ns-hp-gn2-px1122r-l1-l2-rtk-breakout-board/), which requires a UART to USB adapter, which I am using as a RTK GPS Rover with an RPi 4B 4GB.

Regards,
TCIII

**Comment by acredsfan** [comment]
Ok, I actually ordered this bundle: https://a.co/d/ct2dgbk  Thank you for your input on the options, I definitely want to ensure accuracy on this specific component so the extra cost is worth avoiding extra software to achieve acceptable positioning with sub-par equipment.

**Reply by TCIII** [reply]
@acredsfan,
I bought that one initially to build a Base Station and followed this [Sparkfun tutorial](https://learn.sparkfun.com/tutorials/setting-up-a-rover-base-rtk-system). Though I added a [SparkFun Thing Plus - ESP32 WROOM (USB-C)](https://www.amazon.com/gp/product/B0BC29D9QG/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&psc=1) and a [SparkFun 20x4 SerLCD - RGB Backlight (Qwiic)](https://www.amazon.com/gp/product/B08H8XKRDL/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&psc=1) along with this [Sparkfun software](https://github.com/sparkfun/SparkFun_u-blox_GNSS_v3/tree/main/examples/ZED-F9P/Example4_BaseWithLCD) to make putting the ZED-F9P into the Survey-in Mode automatic with the Survey-in status displayed on the LCD Display.

Regards,
TCIII

**Comment by TCIII** [comment]
@acredsfan,

Good news. The latest version of the RC HAT has become available for evaluation and I have verified the functionality of the latest version for the DC RC HAT designer.

Bare RC HAT pwbs should be available in a couple of weeks and I can put one together and send it to you as I don't know how long it will take to make the new RC HAT available to the general public.

The original RC HAT was initially used to decode/encode steering and throttle PWM using the `pigpio` library on the RPi. It would send steering and throttle PWM from the RC Receiver to the RPi for software decoding of the PWM values. It would then receive steering and throttle PWM values from the RPi to control the vehicle steering and throttle over a prerecorded course either using a camera or GPS for input.

Since you are not using RC control of your autonomous mower, you can use the `robohat.py` part to send mower steering and throttle PWM to the RC HAT to provide vehicle direction and speed control. In place of steering and throttle PWM you can send just motor speed PWM to a differential motor controller that can accept two channels of PWM motor control, one for each motor. Or you can use a motor controller that accepts steering and throttle PWM input and creates differential steering and speed control for two attached motors.

Comments please.

Regards,
TCIII

**Comment by TCIII** [comment]
@acredsfan,
Bump?

**Comment by acredsfan** [comment]
@TCIII 
Sorry, I got sidetracked when I was responding the other day and forgot to go back.  I'm excited to add this to the design, as far as the motor controller, I haven't had much experience outside of the IBT-2s and L298Ns that I've used in various iterations of this project.  Are there specific ones you'd recommend for this application?  One of these two?: [Cytron 10A Motor Driver](https://www.cytron.io/p-10amp-5v-30v-dc-motor-driver-2-channels) or [Sabertooth Motor Driver](https://www.robotshop.com/products/sabertooth-dual-2x32a-6v-24v-regenerative-motor-driver?gad_source=1&gclid=CjwKCAjw9cCyBhBzEiwAJTUWNa-Sk6souaEma22PExdPybJ3h4kw42Yk9e612FpQmeoz-JgqfFxj6hoCZgEQAvD_BwE)

Also, let me know what you need from me to send the RC HAT when you're ready and what I owe you for it.  I appreciate you offering to assemble and send one.

Thanks,
acredsfan

**Reply by acredsfan** [reply]
Sorry, one more driver option possibly?  [RoboClaw 2x15A](https://www.robotshop.com/products/roboclaw-2x15a-6-34vdc-regenerative-motor-controller?gad_source=1&gclid=CjwKCAjw9cCyBhBzEiwAJTUWNXawAig18cwjGvoT0MDzX9hixySfLxK499HYVwHeu_1iyLl25Ix_MRoCUcQQAvD_BwE)

**Reply by TCIII** [reply]
@acredsfan,

If you want to use the RC HAT with your RPi 4B, then you will have to use a Motor Driver that accepts RC PWM input which the Cytron 10A Motor Driver you linked above does not, but the Sabertooth Motor Driver does. However, this [Cytron Motor Driver](https://www.cytron.io/p-10amp-7v-30v-dc-motor-driver-for-rc-2-channels) does. I purchased mine from Amazon Prime, but unfortunately it is currently not available.

The RC HAT is not very big or weight very much, so if you cover the postage for first class mail to your address that will be fine with me. I should have two bare RC HAT pwbs, in my possession, to build up in about ten days.

One thing I would like to point out is that setting up your existing autonomous mower for RTK GPS is not trivial, but you have me and my experience available to guide you through the hardware setup process and I think that you will have no problem with the software modifications.

I think that the path forward, once you have the RC motor driver controller, RC HAT, and the RTK GPS hardware available, you need to think about replacing your present `motor_controller.py` with the DC `robohat.py` part and the `gps_interface.py` with the DC `gps.py` part.

Regards,
TCIII

**Reply by acredsfan** [reply]
@TCIII 

This all sounds good, I'll try to start switching over the code and doing some other maintenance to the repository to get ready.  I appreciate all your help on this project!  My username is the same for my Gmail as it is on here, so feel free to email me separately about the RC HAT or anything else. 

Thanks
acredsfan

**Reply by TCIII** [reply]
@acredsfan,
Will do.

Regards,
TCIII

**Comment by TCIII** [comment]
@acredsfan,
Here is a picture of the latest version of the RC HAT.

Regards,
TCIII

![Latest RC HAT](https://github.com/acredsfan/autonomous_mower/assets/1240348/db986a6b-f173-465c-b35a-31065574e4b9)

**Comment by TCIII** [comment]
@acredsfan,

I suggest that you start a branch in your code repository where you can build code that uses the DC `robohat.py` part for motor control, the DC `gps.py` part for your gps interface and a user interface without the sensor inputs and get that code working with the new hardware.

The DC `gps.py` part doesn't care if the GPS input is not RTK GPS and only uses the GNRMC NMEA message for localization. The better the accuracy of the lat and lon values in the GNRMC message the better the localization which is what is achieved when using RTK GPS.

Regards,
TCIII

**Reply by acredsfan** [reply]
Great idea and thank you for the information.  I hope to begin work on that this weekend.

**Reply by TCIII** [reply]
@acredsfan,

Great.

Please post in this discussion thread if you need any help and save the issues thread for documenting new code issues.

Regards,
TCIII

**Comment by tlj1899** [comment]
> Ok, I actually ordered this bundle: https://a.co/d/ct2dgbk  Thank you for your input on the options, I definitely want to ensure accuracy on this specific component so the extra cost is worth avoiding extra software to achieve acceptable positioning with sub-par equipment.
> 
> 

I ordered the same SparkFun GPS-RTK-SMA Kit bundle, scheduled to arrive on Sunday. 

I am still progressing with 3D printing the parts and should have all that printed out and assembled by early next week.

I am now in the process of ordering all the electronics parts.

**Reply by TCIII** [reply]
@tlj1899,

Remember that you will either need a Base Station or access to a NTRIP server to use the Sparkfun ZED-F9P RTK GPS module in the RTK GPS Rover mode otherwise the ZED-F9P RTK GPS is just another expensive, standard GPS.
Here is a [link](https://www.diyrobocars.com/2023/12/28/using-the-donkey-car-path_follow-template-with-rtk-gps/) to how I built a Base Station using the Sparkfun ZED-F9P RTK GPS module and here is a [link](https://learn.sparkfun.com/tutorials/setting-up-a-rover-base-rtk-system) to how to use the ZED-F9P RTK GPS module  with a NTRIP server.

Regards,
TCIII

**Comment by acredsfan** [comment]
@TCIII 

Just an update, I've probably missed some things as I'm trying to keep things as simplistic as possible when integrating, but I've replaced gps_interface.py with gps.py and updated the code in modules that access the gps readings.  I've also updated the README with some of the changes you'd recommended (instructions on obtaining the Google Maps API key, setting up a virtual environment, etc.) and links to newer parts.

Remaining updates I'm working on:
1. Update code using RPi.GPIO (or pigpio) to new methods to work with the RPi 4 boards that are having issues with GPIO communication on Bookworm.
2. Integrate robohat.py to control the wheel motors.
3. General cleanup of old documentation and requirements.txt

I'm also curious what you think about using LIDAR for this project, I would think ambient lighting would make this impractical but was thinking about trying it in a later phase.

Thanks,
acredsfan

**Reply by TCIII** [reply]
@acredsfan,

Excellent news! Please keep up the good work, but remember Family comes first.

I have used the Slamtec A1 and A2 Lidars both experimentally and with ROS on a Arlobot in the past, however there are cheaper Lidars available now. Lidars use IR diodes, but can be affected by external ambient light.
This DC part, [lidar.py,](https://github.com/autorope/donkeycar/blob/main/donkeycar/parts/lidar.py) captures distance and angle, communicates over a USB connection, and might give you and idea of how to integrate a Lidar into your Python code.

Regards,
TCIII

**Comment by TCIII** [comment]
@acredsfan,

I assume that you understand that the DC `gps.py` and `robohat.py` parts will require modification to work with your `main.py` program and other python code within your project code?
I recommend that you create a barebones `main.py` program so that you can easily test the integration of the modified DC `gps.py` and `robohat.py` parts into your python code.

Regards,
TCIII

**Reply by acredsfan** [reply]
@TCIII 

Yes, I've got a list of the modules that the GPS code touches.  As I work on the robohat.py change I'll do the same.  I needed to re work main.py anyway so starting barebones is a good idea. 

Thanks, 
acredsfan

**Reply by acredsfan** [reply]
Actually, on main.py, it's using localization.py to handle tracking the current location and whether the robot is within the yard boundary, but I need to make sure it's all working correctly.  I'm hoping to get some work done on this this weekend so I'll keep you in the loop.

**Reply by TCIII** [reply]
@acredsfan,

Glad to be of help where I can.

Regards,
TCIII

**Reply by acredsfan** [reply]
@TCIII

I appreciate that!  As always, feel free to make any code modifications/improvements you'd like, I'm not protective of any of it, I'm trying to learn and improve as I go and you're helping me with that more than you know.

Thanks, 
acredsfan

**Reply by TCIII** [reply]
@acredsfan,

I can analyze Python code for errors, but I really can't write Python code from scratch. However I have identified several PSF Python Beginner tutorials to help understand the Python structure and coding requirements.

Regards,
TCIII

**Comment by TCIII** [comment]
@acredsfan,

Good news. The latest version of the Donkey Car RC HAT pwb has been received and I will be getting two unpopulated HATs sometime next week.
Once I receive the two HATs I will assemble them and arrange with you to send one your way.

Attached is a photo of the RC HAT pwbs.

Regards,
TCIII

![RC HAT PWBs](https://github.com/acredsfan/autonomous_mower/assets/1240348/f2497c51-7308-4af6-8d22-912da9080ba1)

**Reply by acredsfan** [reply]
Awesome news, @TCIII!  

I'm on vacation in the Tennessee mountains, so I just now saw this.  Will make whatever arrangements necessary when you're ready to ship one my way.

Have a great weekend! 
acredsfan

**Reply by TCIII** [reply]
@acredsfan,

I have just received and tested two of the V0.2 RC HATs so I have one ready to send your way.
The RC HAT does have an OLED display that can be used with the [DC oled.py part](https://github.com/autorope/donkeycar/blob/main/donkeycar/parts/oled.py) on the I2C bus.

This weekend we can figure out by email how to ship the RC HAT to you in Ohio,

Regards,
TCIII

**Reply by acredsfan** [reply]
@TCIII sorry, been out of commission for a s few days but back now.  That's great news, happy to work out the details over email!

Thanks
acredsfan
